Initialise Shared::DAMAGE in PrognosticData::update

einola opened the pull request:
#  Initialise Shared::DAMAGE in PrognosticData::update

Fixes #742 

### Task List
- [x] Defined the tests that specify a complete and functioning change (*It may help to create a [design specification & test specification](../../../wiki/Specification-Template)*)
- [x] Implemented the source code change that satisfies the tests
- [x] Documented the feature by providing worked example
- [x] Updated the README or other documentation
- [x] Completed the pre-Request checklist below

---
# Change Description

The ``Shared::DAMAGE`` variable must be updated at the start of ``PrognosticData::update``, otherwise the ``kernel.setData(damageName, damage)`` call in ``BBMDynamics.cpp`` (line 123) inserts nonsense into the damage field.

---
# Test Description


Use a modified config_benchmark.cfg file to test the update

```
[model]
init_file = init_benchmark_128x128.nc
start = 2023-01-01T00:00:00Z
stop = 2023-01-01T00:10:00Z
time_step = P0-0T00:02:00

[Modules]
DiagnosticOutputModule = Nextsim::ConfigOutput
DynamicsModule = Nextsim::BBMDynamics
IceThermodynamicsModule = Nextsim::DummyIceThermodynamics
LateralIceSpreadModule = Nextsim::DummyIceSpread
AtmosphereBoundaryModule = Nextsim::ConfiguredAtmosphere
OceanBoundaryModule = Nextsim::BenchmarkOcean

[ConfigOutput]
period = P0-0T00:02:00
start = 2010-01-01T12:00:00Z
field_names = hsnow,hice,cice,u,v,damage
filename = benchmark_32x32.diagnostic.nc
```

Before the fix, the damage field would have a uniform slope from x=0 to half way through the domain. After the fix, the damage field is uniform as it should be.

---
# Documentation Impact

N/A

---
# Other Details

N/A

---
### Pre-Request Checklist

- [x] The requirements of this pull request are fully captured in an issue or design specification and are linked and summarised in the description of this PR
- [x] No new warnings are generated
- [x] The documentation has been updated (or an issue has been created to track the corresponding change)
- [x] Methods and Tests are commented such that they can be understood without having to obtain additional context
- [x] This PR/Issue is labelled as a bug/feature/enhancement/breaking change
- [x] File dates have been updated to reflect modification date
- [x] This change conforms to the conventions described in the README

einola committed:
Initialise Shared::DAMAGE in PrognosticData::update

The Shared::DAMAGE variable needs to be updated at the start of
PrognosticData::update, otherwise the kernel.setData(damageName, damage)
 in BBMDynamics.cpp (line 123) inserts nonsense into the damage field.

timspainNERSC commented:
An alternative suggestion, so that `damage` works like all the other fields: 
In `ConstantHealing.hpp` add:
```
    ModelArrayRef<Protected::DAMAGE> oldDamage;
```
In `ConstantHealing.cpp` change
```
    damage[i] = (damage[i] * (cice[i] - lateralGrowth) + lateralGrowth) / cice[i];
```
to 
```
    damage[i] = (oldDamage[i] * (cice[i] - lateralGrowth) + lateralGrowth) / cice[i];
```
But still delete the reference to `ticeUpd` in `PrognosticData.cpp`.

einola commented:
Yes, that makes more sense. It is at least more in the spirit of the thing. But then I also have to do ``damage = oldDamage`` in ``NoHealing``.

einola commented:
I'll do that and then get to  #725 sometime later.

einola committed:
Change where we update the main damage variable

Following Tim's suggestion, the update is now done in the DamageHealing
module.

einola commented:
With 

```
    ModelArrayRef<Shared::DAMAGE, RW> damage; 
    ModelArrayRef<Protected::DAMAGE, RW> oldDamage;
```

right? It crashes with a segfault/bad access, so I'm probably misunderstanding something :/

timspainNERSC commented:
> ModelArrayRef<Protected::DAMAGE, RW> oldDamage;

The array at `Protected::DAMAGE` is shared read-only, so attempting to access it read-write causes an access violation. The error message could be a little more subtle, but that's how it currently is.
```
ModelArrayRef<Protected::DAMAGE, RO> oldDamage;
```

einola commented:
Yes, I realised just after I posted :)

timspainNERSC commented on physics/src/modules/DamageHealingModule/include/NoHealing.hpp:
```diff
@@ -26,7 +26,14 @@ class NoHealing : public IDamageHealing {
 
     // Do nothing when update is called.
-    void update(const TimestepTime& tstep) override {};
+    void update(const TimestepTime& tstep) override
```
`damage.data() = oldDamage.data();` should work without the `overElements()` call.

einola committed:
Simplify NoHealing

Use damage.data() = oldDamage.data() instead of overElements

einola committed:
Update DamageHealing_test.cpp to use oldDamage

The test didn't work because oldDamage wasn't set.

einola committed:
Minor changes to DamageHealing_test.cpp

Protected::DAMAGE should be RO and Module.hpp is not needed (and
confuses the compiler if it's there).

einola committed:
Update to IceGrowth_test.cpp to reflect new damage handling

IceGrowth_test.cpp also has to define both damage and oldDamage now.
